Multicoloured spesos: add layer function in StackComponent

whatston3 opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->

Spesos should now appear in non-single denominations.

StackComponents now have an optional layerFunction that will adjust the layer data before it is used in the StackVisualizer.

This is only used in spesos stacks at the moment, with one concrete instance that uses threshold values (used with bill denominations for the purpose of spesos).

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->

Looks pretty, feels better when you make more money.

My biggest concerns here are whether this is fragile with respect to upstream changes, and if the names of things make sense.

## How to test
<!-- Describe the way it can be tested -->

Build and run local server.
1. Spawn in spesos in $1, $10, $100, $500, $1000, and $1000000 denominations.
2. Try merging and splitting 1 off from all bills larger than $1.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

![spesos](https://github.com/new-frontiers-14/frontier-station-14/assets/166147148/e52cd713-e79f-4e6f-a9fc-b39c340790e2)

- [X] ***I have added screenshots/videos to this PR showcasing its changes ingame***, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl:
- fix: Stacks of spesos now display the largest valued bill on top.

dvir001 commented:
Can you also cap the stack to max $1000000 / 100000 ?
Might as well since its a bit weird you can have infinite amounts on a single stack

whatston3 commented:
> Can you also cap the stack to max $1000000 / 100000 ? Might as well since its a bit weird you can have infinite amounts on a single stack

I'll look into how this works with cargo sales and ATM withdrawals, not the largest fan of the idea, makes handling large transfers (e.g. spessman's lotto) awkward.

Leander-0 commented:
I like the idea of the money being stacked, we can add money bags that act as the mining bags to handle large amounts of money. kinda silly that one can carry 1 million spessos on their pocket without issue. also adds gameplay to hide the money from pirates.

whatston3 commented:
> I'll look into how this works with cargo sales and ATM withdrawals, not the largest fan of the idea, makes handling large transfers (e.g. spessman's lotto) awkward.

Gonna say this is out of the scope of the PR.  Capping the max stack size causes issues with ATM withdrawal ("admin, help, I asked for 5 million spesos, ATM gave me 1 and ate the other 4").  What I'd argue _is_ in scope is a few new bill denominations (10K, 25K, 50K, and 100K all scroll - space on bill face for amount is tiny - but do not change colour).  Pink bill needs rework, my spriteing is awful.

![image](https://github.com/new-frontiers-14/frontier-station-14/assets/166147148/c60f0894-0a50-4542-9a77-15f6ae5f7f2b)

dvir001 commented:
> > I'll look into how this works with cargo sales and ATM withdrawals, not the largest fan of the idea, makes handling large transfers (e.g. spessman's lotto) awkward.
> 
> Gonna say this is out of the scope of the PR. Capping the max stack size causes issues with ATM withdrawal ("admin, help, I asked for 5 million spesos, ATM gave me 1 and ate the other 4"). What I'd argue _is_ in scope is a few new bill denominations (10K, 25K, 50K, and 100K all scroll - space on bill face for amount is tiny - but do not change colour). Pink bill needs rework, my spriteing is awful.
> 
> ![image](https://private-user-images.githubusercontent.com/166147148/338288143-c60f0894-0a50-4542-9a77-15f6ae5f7f2b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgwNTg3NjAsIm5iZiI6MTcxODA1ODQ2MCwicGF0aCI6Ii8xNjYxNDcxNDgvMzM4Mjg4MTQzLWM2MGYwODk0LTBhNTAtNDU0Mi05YTc3LTE1ZjZhZTVmN2YyYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxMFQyMjI3NDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYjBjNjdjMjcwODM1M2MyYzhkYjA3OWM2OWE0ODI5ZGRlYzc1NWY4YmM4OGZmMDY3NjNiNTg3NWZlMDMyNTFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.5yodm3-I0DHJicHyQ6CcZ9uE3mOyrhTCqkLbKbMMG4w)

Sound good

whatston3 commented:
Adjusted style of large bills:
* all scroll left
* all use 4-pixel wide zeros
* colours avoid repetition as best as possible
* pink bill changed to a bright red [still probably too bright]
* rainbow bill flashing follows order of currency, uses buff colour at bottom of digits
* rainbow bill value changed to 250k (rationale here: if we will have a maximum size at some point, you shouldn't be able to hide your flashing bills easily by splitting it easily

Preview of bills as they stand (bottom row are animated - 10k, 25k, 50k, 100k, 250k+)
![image](https://github.com/new-frontiers-14/frontier-station-14/assets/166147148/4ce97087-4046-41c9-9d67-a8d6ba4fac5b)

Cheackraze commented on Content.Shared/Stacks/StackComponent.cs:
```diff
@@ -78,6 +78,14 @@ public sealed partial class StackComponent : Component
+        /// An optional function to adjust the layers used for a stack's appearance.
+        /// </summary>
+        [DataField("layerFunction")]
+        public StackLayerFunction? LayerFunction;
```
hm this seems really smelly to me. It seems like you're storing a function into a component, which isnt serializable and will break any kind of saving/loading or serialization. The benefits of keeping components to only data, and a fully ECS system.

Cheackraze commented on Content.Client/Stack/StackSystem.cs:
```diff
@@ -56,20 +56,25 @@ private void OnAppearanceChange(EntityUid uid, StackComponent comp, ref Appearan
-            if (!_appearanceSystem.TryGetData<bool>(uid, StackVisuals.Hide, out var hidden, args.Component))
-                hidden = false;
+            if (comp.LayerFunction is not null) // Frontier: use stack layer function to modify appearance if provided.
+                comp.LayerFunction.Apply(ref data); // Frontier
```
instead, you should just add a bool flag to the stack component to check for that instead, and then call the shared function here after bringing it in as a dependency.

whatston3 commented:
Will review this with fresher eyes before requesting another review, but @Cheackraze, your concerns should be addressed.

whatston3 commented:
Changes should be addressed.  Still functional, and if you have a better name for `ThresholdStackLayerFunctionComponent`, I'm all ears.

Cheackraze commented:
> Changes should be addressed. Still functional, and if you have a better name for `ThresholdStackLayerFunctionComponent`, I'm all ears.

at is a bit wordy for what it is. Its just a component, its use for a function is irrelevant. maybe just, 'StackLayerThresholdComponent'

whatston3 commented:
> at is a bit wordy for what it is. Its just a component, its use for a function is irrelevant. maybe just, 'StackLayerThresholdComponent'

Sounds good to me.  Renamed file, reworded a comment, adjusted component name in speso YAML definition.  Still works alright, no issues.
